File: e2e/runner/main.go
// E2E test runner — cross-platform Go binary.
//
// 1. Builds openerpd and openerp CLI via Bazel
// 2. Creates a test context using `openerp context create --password`
// 3. Starts openerpd server on a random port
// 4. Waits for /health to return 200
// 5. Runs Rust unit tests
// 6. Runs Node.js E2E tests
// 7. Kills server, cleans up
package main

import (
	"fmt"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

const rootPass = "openerp123"

func main() {
	root := findWorkspaceRoot()
	if root == "" {
		fatal("cannot find workspace root (MODULE.bazel)")
	}
	fmt.Println("Workspace:", root)

	// Step 1: Build both binaries.
	fmt.Println("\n=== Step 1: Build binaries ===")
	if err := bazel(root, "build", "//rust/bin/openerpd", "//rust/bin/openerp"); err != nil {
		fatal("Build failed: %v", err)
	}

	ext := ""
	if runtime.GOOS == "windows" {
		ext = ".exe"
	}
	openerpd := filepath.Join(root, "bazel-bin/rust/bin/openerpd/openerpd"+ext)
	openerp := filepath.Join(root, "bazel-bin/rust/bin/openerp/openerp"+ext)

	// Step 2: Run Rust tests.
	fmt.Println("\n=== Step 2: Rust tests ===")
	rustTests := []string{
		"//rust/lib/dsl/golden:golden_test",
		"//rust/lib/dsl/store:store_test",
		"//rust/lib/dsl/types:types_test",
		"//rust/lib/dsl/macro_test:macro_test",
		"//rust/lib/core:core_test",
		"//rust/mod/auth:auth_test",
	}
	if err := bazel(root, append([]string{"test"}, rustTests...)...); err != nil {
		fatal("Rust tests failed: %v", err)
	}
	fmt.Println("Rust tests passed.")

	// Step 3: Create test context via CLI.
	fmt.Println("\n=== Step 3: Create test context ===")
	tmpDir, err := os.MkdirTemp("", "openerp-e2e-*")
	if err != nil {
		fatal("temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	configDir := filepath.Join(tmpDir, "config")
	dataDir := filepath.Join(tmpDir, "data")
	clientConfig := filepath.Join(tmpDir, "client.toml")

	cmd := exec.Command(openerp,
		"--config", clientConfig,
		"context", "create", "e2e-test",
		"--config-dir", configDir,
		"--data-dir", dataDir,
		"--password", rootPass,
	)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		fatal("context create: %v", err)
	}

	// Find the generated server config.
	serverConfig := filepath.Join(configDir, "e2e-test.toml")
	if _, err := os.Stat(serverConfig); os.IsNotExist(err) {
		fatal("server config not found at %s", serverConfig)
	}
	fmt.Printf("Server config: %s\n", serverConfig)

	// Step 4: Start server on random port.
	fmt.Println("\n=== Step 4: Start server ===")
	port := freePort()
	listen := fmt.Sprintf("127.0.0.1:%d", port)
	baseURL := fmt.Sprintf("http://%s", listen)

	srv := exec.Command(openerpd, "-c", serverConfig, "--listen", listen)
	srv.Env = append(os.Environ(), "RUST_LOG=warn")
	srv.Stdout = os.Stdout
	srv.Stderr = os.Stderr
	if err := srv.Start(); err != nil {
		fatal("start server: %v", err)
	}
	defer func() {
		srv.Process.Kill()
		srv.Wait()
	}()

	if !waitForHealth(baseURL, 30*time.Second) {
		fatal("server did not start within 30s")
	}
	fmt.Printf("Server running on %s\n", baseURL)

	// Step 5: Install E2E deps if needed.
	fmt.Println("\n=== Step 5: Install E2E deps ===")
	e2eDir := filepath.Join(root, "e2e")
	if _, err := os.Stat(filepath.Join(e2eDir, "node_modules")); os.IsNotExist(err) {
		npm := exec.Command("npm", "install")
		npm.Dir = e2eDir
		npm.Env = append(os.Environ(), "PUPPETEER_SKIP_DOWNLOAD=true")
		npm.Stdout = os.Stdout
		npm.Stderr = os.Stderr
		if err := npm.Run(); err != nil {
			fatal("npm install: %v", err)
		}
	}

	// Step 6: Run E2E tests.
	fmt.Println("\n=== Step 6: Run E2E tests ===")
	testFiles := []string{
		"tests/01-login.test.mjs",
		"tests/02-dashboard-crud.test.mjs",
		"tests/03-api-auth.test.mjs",
		"tests/04-user-login.test.mjs",
		"tests/05-facet-api.test.mjs",
	}
	args := append([]string{"--test"}, testFiles...)
	node := exec.Command("node", args...)
	node.Dir = e2eDir
	node.Env = append(os.Environ(),
		"BASE_URL="+baseURL,
		"ROOT_PASS="+rootPass,
	)
	node.Stdout = os.Stdout
	node.Stderr = os.Stderr
	if err := node.Run(); err != nil {
		fatal("E2E tests failed: %v", err)
	}

	fmt.Println("\n=== All tests passed! ===")
}

func findWorkspaceRoot() string {
	dir, _ := os.Getwd()
	for {
		if _, err := os.Stat(filepath.Join(dir, "MODULE.bazel")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return ""
		}
		dir = parent
	}
}

func bazel(dir string, args ...string) error {
	cmd := exec.Command("bazel", args...)
	cmd.Dir = dir
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

func freePort() int {
	l, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		fatal("free port: %v", err)
	}
	port := l.Addr().(*net.TCPAddr).Port
	l.Close()
	return port
}

func waitForHealth(baseURL string, timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		resp, err := http.Get(baseURL + "/health")
		if err == nil && resp.StatusCode == 200 {
			resp.Body.Close()
			return true
		}
		time.Sleep(500 * time.Millisecond)
	}
	return false
}

func fatal(format string, args ...interface{}) {
	msg := fmt.Sprintf(format, args...)
	if !strings.HasSuffix(msg, "\n") {
		msg += "\n"
	}
	fmt.Fprintf(os.Stderr, "FATAL: %s", msg)
	os.Exit(1)
}

File: e2e/runner/src/main.rs
//! E2E test runner — cross-platform Rust binary.
//!
//! 1. Builds openerpd + openerp CLI via Bazel
//! 2. Creates a test context using `openerp context create --password`
//! 3. Starts openerpd on a random port
//! 4. Waits for /health to return 200
//! 5. Runs Rust unit tests via Bazel
//! 6. Runs Node.js E2E tests
//! 7. Kills server, cleans up

use std::io::{BufRead, BufReader};
use std::net::TcpListener;
use std::path::{Path, PathBuf};
use std::process::{Child, Command, Stdio};
use std::time::{Duration, Instant};

const ROOT_PASS: &str = "openerp123";

fn main() {
    let root = find_workspace_root().unwrap_or_else(|| {
        fatal("Cannot find workspace root (MODULE.bazel). Run via: bazel run //e2e/runner");
    });
    // When launched by `bazel run`, cwd is the sandbox. Switch to real workspace.
    std::env::set_current_dir(&root).expect("chdir to workspace");
    println!("Workspace: {}", root.display());

    // Step 1: Build binaries.
    step("Step 1: Build binaries");
    bazel(&root, &["build", "//rust/bin/openerpd", "//rust/bin/openerp"]);

    let ext = if cfg!(windows) { ".exe" } else { "" };
    let openerpd = root.join(format!("bazel-bin/rust/bin/openerpd/openerpd{ext}"));
    let openerp = root.join(format!("bazel-bin/rust/bin/openerp/openerp{ext}"));

    // Step 2: Rust unit tests.
    step("Step 2: Rust tests");
    bazel(
        &root,
        &[
            "test",
            "//rust/lib/dsl/golden:golden_test",
            "//rust/lib/dsl/store:store_test",
            "//rust/lib/dsl/types:types_test",
            "//rust/lib/dsl/macro_test:macro_test",
            "//rust/lib/core:core_test",
            "//rust/mod/auth:auth_test",
            "//rust/mod/pms:pms_test",
            "//rust/mod/task:task_test",
        ],
    );
    println!("Rust tests passed.");

    // Step 3: Create test context via CLI.
    step("Step 3: Create test context");
    let tmp_dir = tempfile::tempdir().expect("create temp dir");
    let config_dir = tmp_dir.path().join("config");
    let data_dir = tmp_dir.path().join("data");
    let client_config = tmp_dir.path().join("client.toml");

    run(
        &openerp,
        &[
            "--config",
            client_config.to_str().unwrap(),
            "context",
            "create",
            "e2e-test",
            "--config-dir",
            config_dir.to_str().unwrap(),
            "--data-dir",
            data_dir.to_str().unwrap(),
            "--password",
            ROOT_PASS,
        ],
    );

    let server_config = config_dir.join("e2e-test.toml");
    assert!(
        server_config.exists(),
        "Server config not found: {}",
        server_config.display()
    );
    println!("Server config: {}", server_config.display());

    // Step 4: Start server on a random port.
    step("Step 4: Start server");
    let port = free_port();
    let listen = format!("127.0.0.1:{port}");
    let base_url = format!("http://{listen}");

    let mut server = Command::new(&openerpd)
        .args(["-c", server_config.to_str().unwrap(), "--listen", &listen])
        .env("RUST_LOG", "warn")
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .expect("start openerpd");

    // Stream server stderr in a background thread so we can see logs.
    let stderr = server.stderr.take().unwrap();
    let log_thread = std::thread::spawn(move || {
        let reader = BufReader::new(stderr);
        for line in reader.lines() {
            if let Ok(line) = line {
                eprintln!("[openerpd] {line}");
            }
        }
    });

    let _guard = ServerGuard {
        child: server,
        _log_thread: log_thread,
    };

    wait_for_health(&base_url, Duration::from_secs(30));
    println!("Server running on {base_url}");

    // Step 5: Install E2E Node deps if needed.
    step("Step 5: Install E2E deps");
    let e2e_dir = root.join("e2e");
    if !e2e_dir.join("node_modules").exists() {
        let mut cmd = Command::new("npm");
        cmd.arg("install")
            .current_dir(&e2e_dir)
            .env("PUPPETEER_SKIP_DOWNLOAD", "true");
        let status = cmd.status().expect("npm install");
        if !status.success() {
            fatal("npm install failed");
        }
    }

    // Step 6: Run E2E tests.
    step("Step 6: Run E2E tests");
    let test_files = [
        "tests/01-login.test.mjs",
        "tests/02-dashboard-crud.test.mjs",
        "tests/03-api-auth.test.mjs",
        "tests/04-user-login.test.mjs",
        "tests/05-facet-api.test.mjs",
        "tests/06-pms-actions.test.mjs",
        "tests/07-task-actions.test.mjs",
        "tests/08-put-edit-api.test.mjs",
        "tests/09-user-password-login.test.mjs",
    ];
    let mut args: Vec<&str> = vec!["--test"];
    args.extend(test_files.iter());

    let status = Command::new("node")
        .args(&args)
        .current_dir(&e2e_dir)
        .env("BASE_URL", &base_url)
        .env("ROOT_PASS", ROOT_PASS)
        .status()
        .expect("run node tests");
    if !status.success() {
        fatal("E2E tests failed");
    }

    println!("\n=== All tests passed! ===");
}

// ── Helpers ──

fn find_workspace_root() -> Option<PathBuf> {
    // Prefer BUILD_WORKSPACE_DIRECTORY (set by `bazel run`).
    if let Ok(dir) = std::env::var("BUILD_WORKSPACE_DIRECTORY") {
        let p = PathBuf::from(&dir);
        if p.join("MODULE.bazel").exists() {
            return Some(p);
        }
    }
    // Fallback: walk up from cwd.
    let mut dir = std::env::current_dir().ok()?;
    loop {
        if dir.join("MODULE.bazel").exists() {
            return Some(dir);
        }
        if !dir.pop() {
            return None;
        }
    }
}

fn step(name: &str) {
    println!("\n=== {name} ===");
}

fn bazel(dir: &Path, args: &[&str]) {
    let status = Command::new("bazel")
        .args(args)
        .current_dir(dir)
        .status()
        .expect("run bazel");
    if !status.success() {
        fatal(&format!("bazel {} failed", args.join(" ")));
    }
}

fn run(bin: &Path, args: &[&str]) {
    let status = Command::new(bin)
        .args(args)
        .status()
        .unwrap_or_else(|e| fatal(&format!("run {}: {e}", bin.display())));
    if !status.success() {
        fatal(&format!("{} failed", bin.display()));
    }
}

fn free_port() -> u16 {
    let listener = TcpListener::bind("127.0.0.1:0").expect("bind free port");
    listener.local_addr().unwrap().port()
}

fn wait_for_health(base_url: &str, timeout: Duration) {
    let url = format!("{base_url}/health");
    let deadline = Instant::now() + timeout;
    while Instant::now() < deadline {
        if let Ok(resp) = reqwest::blocking::get(&url) {
            if resp.status().is_success() {
                return;
            }
        }
        std::thread::sleep(Duration::from_millis(500));
    }
    fatal(&format!(
        "Server did not become healthy within {}s",
        timeout.as_secs()
    ));
}

fn fatal(msg: &str) -> ! {
    eprintln!("FATAL: {msg}");
    std::process::exit(1);
}

/// RAII guard — kills the server process on drop.
struct ServerGuard {
    child: Child,
    _log_thread: std::thread::JoinHandle<()>,
}

impl Drop for ServerGuard {
    fn drop(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}
